### ISSUE TITLE
Default build process may crash users machine

### ISSUE THREAD

**Issue by hasufell** [haskell/haskell-ide-engine#1302]
Just tried to build HIE and it spiked my memory up, bombed my swap and made my laptop unresponsive for straight 20 minutes before I could even abort the build process, endangering my files and current work.

This is one of the worst things with haskell build tools. Making assumptions about users memory in a language where the garbage collector is totally unpredictable.

Related:
* https://github.com/haskell/cabal/issues/5776
* https://github.com/commercialhaskell/stack/issues/4296

**Comment by Anrock** [comment]
Sounds like upstream issue to me. Not sure what is expected action here.

**Comment by hasufell** [comment]
> Sounds like upstream issue to me. Not sure what is expected action here.

Run cabal with `-j1`.

**Comment by lorenzo** [comment]
I think the best way forward for hie is to offer binaries, instead of asking users to compile it from scratch.

**Comment by hasufell** [comment]
> I think the best way forward for hie is to offer binaries, instead of asking users to compile it from scratch.

That's a totally different issue. And I do not want to use binaries.

**Comment by fendor** [comment]
Cant we also try to restrict the memory? 
The flag `-j1` instructs shake to only use a single core for the build environment, e.g. shake rules. We can define additional flags that are passed to stack/cabal to enforce a specific amount of cores.

**Comment by lorenzo** [comment]
> That's a totally different issue. And I do not want to use binaries

Cool, but I still think that’s the best way forward to shield most users from all the installation problems that’s may arise, such as this one. You, as an advanced user who does not want binaries, can still tweak the configuration.

**Comment by Anrock** [comment]
@hasufell were you building HIE via Stack? If so, wouldn't it be easier to adjust number of build threads in your global stack config? 

I, personally, never had problems with builds locking up my system on any of my machines (one is octacore with 16GB and second is quadrocore with same RAM) and OSes (Lin, Win, OSX), so to me it sounds like this issues is specific to your machine.

**Comment by hasufell** [comment]
> @hasufell were you building HIE via Stack? If so, wouldn't it be easier to adjust number of build threads in your global stack config?

`stack install.hs cabal-hie-8.6.5`

> so to me it sounds like this issues is specific to your machine.

I don't think so. Other users have reported similar problems. HIE is especially difficult, because it requires you to go through shake.

**Comment by MoHKale** [comment]
I agree with @lorenzo. HIE has way too many dependencies and takes far too long to build for it to be practical to build it yourself on windows, where it failing to build for a completely arbitrary reason isn't too out of the ordinary. I tried to install it 9 times yesterday over the course of 6 hours. By the end I still couldn't get it running. I gave up and intend to try again today.
